src/software/new_geom/cubic_bezier_spline2d.cpp

```diff
@@ -0,0 +1,167 @@
+    // the first segment is t=0, the end of the last segment is t=1, and the
+    // the knots are evenly distributed in between 0 and 1
+
+
+    const double t_constrained = std::clamp(t, 0.0, 1.0);
+
+    // Scale t from [0,1] -> [0,n] where n is the number segments
+    const double t_mapped = t_constrained * static_cast<double>(getNumSegments());
+
+    // In most cases the segment index is just the floor of the mapped t, but we
+    // also need to have extra handling for the case where t=1, in which case
+    // we need to just use the last segment
+    const size_t segment_index =
+        std::min(static_cast<size_t>(std::floor(t_mapped)), getNumSegments() - 1);
+
+    // Figure out what the t-value on the segment is going to be
+    const double t_on_segment = (t_mapped - static_cast<double>(segment_index));
+
+    // Generate the bezier curve for the segment and evaluate it at the segment t-value
+    const BezierCurve2d segment = getSegmentAtIndex(segment_index);
+    return segment.getValueAt(t_on_segment);
+}
+
+const std::vector<Point> CubicBezierSpline2d::getKnots() const
+{
+    std::vector<Point> knots;
+    for (size_t i = 0; i < control_points.size(); i += 3)
+    {
+        knots.emplace_back(control_points[i]);
+    }
+
+    return knots;
+}
+
+const std::vector<Point>& CubicBezierSpline2d::getControlPoints() const
+{
+    return control_points;
+}
+
+size_t CubicBezierSpline2d::getNumKnots() const
+{
+    return static_cast<size_t>((control_points.size() - 1) / 3 + 1);
+}
+
+size_t CubicBezierSpline2d::getNumSegments() const
+{
+    return getNumKnots() - 1;
+}
+
+
+std::vector<double> CubicBezierSpline2d::getKnotParametrizationValues() const
+{
+    // We assume a linear spacing of all the knots from 0 to 1
+    std::vector<double> knot_vector;
+    const double num_knots = getNumKnots();
+    for (size_t i = 0; i < num_knots; i++)
+    {
+        knot_vector.emplace_back(i * 1.0 / static_cast<double>(num_knots - 1));
+    }
+    return knot_vector;
+}
+
+const Point CubicBezierSpline2d::getStartPoint() const
+{
+    // We are guaranteed to at least have two control points, the start and end points
+    return control_points.front();
+}
+
+const Point CubicBezierSpline2d::getEndPoint() const
+{
+    // We are guaranteed to at least have two control points, the start and end points
+    return control_points.back();
+}
+
+const std::vector<SplineSegment2d> CubicBezierSpline2d::getSplineSegments() const
+{
+    std::vector<SplineSegment2d> segments;
+    const double num_knots = getNumKnots();
+    for (size_t i = 0; i < num_knots - 1; i++)
+    {
+        const BezierCurve2d curve = getSegmentAtIndex(i);
+        segments.emplace_back(createSplineSegment2d(0, 1, curve.getPolynomial()));
+    }
+    return segments;
+}
+
+BezierCurve2d CubicBezierSpline2d::getSegmentAtIndex(size_t index) const
+{
+    return BezierCurve2d({control_points[index * 3], control_points[index * 3 + 1],
```

Comment:
since we're using `3` so much perhaps its worth making a class-constant like `DEGREE_OF_SPLINE` (or something better) since it is related to the cubic nature of the representation. Not sure if that would make it more readable.

Reply:
Hmm, I'd be inclined not to? I'm just not sure what would be a reasonable name for it and if it would actually improve things much.

Reply:
yea I'm fine with it as-is